packages/tpetra/core/src/Tpetra_Import_def.hpp

```diff
@@ -830,466 +945,105 @@ namespace Tpetra {
-    comm->barrier ();
-    {
-      std::ostringstream os;
-      os << myRank << ": srcGIDs: " << toString (srcGIDs) << endl;
-      os << myRank << ": tgtGIDs1: " << toString (tgtGIDs1) << endl;
-      os << myRank << ": tgtGIDs2: " << toString (tgtGIDs2) << endl;
-      cerr << os.str ();
+    // Create a mapping from remote GID -> PID.  The mapping is used to sort
+    // remote GIDs in the union target map by PID.
+    ArrayView<const LO> remoteLIDs1 = this->getRemoteLIDs();
+    Array<int> remotePIDs1;
+    Tpetra::Import_Util::getRemotePIDs(*this, remotePIDs1);
+    std::map<GO,int> remoteGIDProc;
+    for (size_type k=0; k < remoteLIDs1.size(); k++) {
+      GO GID = this->getTargetMap()->getGlobalElement(remoteLIDs1[k]);
+      remoteGIDProc[GID] = remotePIDs1[k];
     }
-    comm->barrier ();
-#endif // HAVE_TPETRA_IMPORT_SETUNION_EXTRA_DEBUG_OUTPUT
-
 
-    // Fill this as we go with the union target Map's GIDs, in the
-    // desired order.  We'll need them for the Map constructor.
-    Array<GO> unionTgtGIDs;
-    // Upper bound on the number of union target Map GIDs.  This
-    // happens to be strict, but doesn't have to be.  Setting some
-    // reasonable upper bound avoids reallocation in loops that do
-    // push_back operations.
-    unionTgtGIDs.reserve (tgtGIDs1.size () + tgtGIDs2.size ());
-
-#ifdef HAVE_TPETRA_IMPORT_SETUNION_EXTRA_DEBUG_OUTPUT
-    if (myRank == 0) {
-      cerr << endl;
+    ArrayView<const LO> remoteLIDs2 = rhs.getRemoteLIDs();
+    Array<int> remotePIDs2;
+    Tpetra::Import_Util::getRemotePIDs(rhs, remotePIDs2);
+    for (size_type k=0; k < remoteLIDs2.size(); k++) {
+      GO GID = rhs.getTargetMap()->getGlobalElement(remoteLIDs2[k]);
+      // If GID is already in remoteGIDProc it will just be overwritten which
+      // okay since the PID will be the same.
+      if (debug) {
+        auto iter = remoteGIDProc.find(GID);
+        if (iter != remoteGIDProc.end()) {
+          // Remote GID already registered in map
+          TEUCHOS_TEST_FOR_EXCEPTION(iter->second!=remotePIDs2[k],
+              std::logic_error,
+              "Expected PID for remote GID " << GID <<
+              " to be " << iter->second <<
+              " got " << remotePIDs2[k] << " instead.");
+        }
+      }
+      remoteGIDProc[GID] = remotePIDs2[k];
     }
-    comm->barrier ();
-    comm->barrier ();
-    comm->barrier ();
-    cerr << myRank << ": Computing \"same\" GIDs" << endl;
-#endif // HAVE_TPETRA_IMPORT_SETUNION_EXTRA_DEBUG_OUTPUT
-
-    // Compute the initial sequence of "same" GIDs in the union
-    // import.  The number of "same" GIDs in the union is the maximum
-    // of the lengths of this in the two inputs.
 
+    // Get the union of the target GIDs
+    Array<GO> unionTgtGIDs;
+    size_type numSameIDsUnion, numPermuteIDsUnion, numRemoteIDsUnion;
     const size_type numSameGIDs1 = this->getNumSameIDs ();
+    ArrayView<const GO> tgtGIDs1 = tgtMap1->getNodeElementList ();
```

Comment:
In the future, we'll want to move away from calling `getNodeElementList`, because it may require copying from device to host, and because it forces contiguous Maps to allocate an array of GIDs.  This is fine for now, though.